### PR TITLE
feat (3.4.0): add platform option, Android and BSD platforms

### LINKED ISSUE #395: Rendered emails become URL-encoded
## Actual behavior

When the TLDR is rendered, emails are URL-encoded. For example using `tldr gpg`:

```
  - Export public key for mailto:&#x61;&#108;&#x69;&#99;&#x65;&#x40;&#x65;&#x78;&#x61;&#x6d;&#112;&#x6c;&#x65;&#x2e;&#x63;&#111;&#x6d; (output to stdout):
    gpg --export --armor alice@example.com
```

## Expected behavior

Using the Python client and the same example, the output is correct:

```
  - Export public key for alice@example.com (output to stdout):
    gpg --export --armor alice@example.com
```

## Environment

- Node.js v18.12.1
- tldr v3.3.8

Accidentally made [this issue in tldr](https://github.com/tldr-pages/tldr/issues/9560) first. (^:

### PR DESCRIPTION
**Pull request by kbdharun**
## Description

This PR updates the `--os` and `-o` option to `--platform` and `-p`. This change would make the official client comply with the last client specification and #420 is expected to add support for the recent client specification 2.0. (Supersedes #337, thanks to the original authors and reviews for initial work on this transition)

This PR also adds support for Android and other BSD platforms (~there might be another platform `openbsd` in future, subscribe to https://github.com/tldr-pages/tldr/pull/10698 to stay tuned~ Edit. added it too).

## Checklist

Please review this checklist before submitting a pull request.

- [x] Code compiles correctly
- [x] Created tests, if possible
- [x] All tests passing (`npm run test:all`)
- [x] Extended the README / documentation, if necessary

## Related PRs

Closes #337
Fixes #395

### CODE REVIEW

File: bin/tldr
Comment (MasterOdin) [inline bin/tldr:25]: From the other issue, I don't think it's worth fully dropping `-o` and `--os` given that not using either currently, and instead would add it as an [`.alias`](https://github.com/tj/commander.js/blob/master/examples/alias.js). It won't be documented, but if people still use it, :shrug:

Could maybe look to see if could add a way to detect if `--os` was used vs `--platform` and show a deprecation warning to eventually remove it.

File: bin/tldr
Comment (MasterOdin) [inline bin/tldr:66]: Match style of other code:
```suggestion
if(program.android) {
```

File: bin/tldr
Comment (kbdharun) [reply bin/tldr:25]: Thanks for the suggestion, will try out this alias method instead. (Not really familiar with NodeJS but will figure it out, if I have any doubts will reach out)

File: bin/tldr
Comment (MasterOdin) [inline bin/tldr:7]: I think the discussion in #337 got this wrong, where naming this `Platform` is against standard conventions, where if it's capitalized, then it's assumed that we're referring to a class object, vs a regular object. I'd personally say rename the file to `lib/platforms` and have the import be `const platforms = require('../lib/platforms')`.

File: lib/platform.js
Comment (MasterOdin) [inline lib/platform.js:1]: I would rename this file to be `platforms.js` to match the import variable we use elsewhere.

**Comment by kbdharun** [comment]
> Any update on this?

I am planning on adding support for openbsd platform too in this, will do this later today, along with another change.

File: bin/tldr
Comment (MasterOdin) [inline bin/tldr:93]:
```suggestion
if (program.platform && platforms.isSupported(program.platform)) {
  cfg.platform = program.platform;
}
```

**Review by MasterOdin — APPROVED** [review]
LGTM to me.

We could simplify adding new platforms by doing something like this:

```diff
diff --git a/bin/tldr b/bin/tldr
index 59e0b91..21e28ba 100755
--- a/bin/tldr
+++ b/bin/tldr
@@ -22,12 +22,11 @@ program
   .option('-e, --random-example', 'Show a random example')
   .option('-f, --render [file]', 'Render a specific markdown [file]')
   .option('-m, --markdown', 'Output in markdown format')
-  .option('-p, --platform [type]', 'Override the current platform [android, linux, osx, sunos, windows]')
-  .option('--android', 'Override the platform with Android')
-  .option('--linux', 'Override the platform with Linux')
-  .option('--osx', 'Override the platform with OSX')
-  .option('--sunos', 'Override the platform with SunOS')
-  .option('--windows', 'Override the platform with Windows')
+  .option('-p, --platform [type]', `Override the current platform [${platforms.supportedPlatforms.join(', ')}]`);
+for (const platform of platforms.supportedPlatforms) {
+  program.option(`--${platform}`, `Override the platform with ${platform}`);
+}
+program
   .alias('-o', '-p') // Alias -o to -p
   .alias('--os', '--platform') // Alias --os to --platform
   .option('-t, --theme [theme]', 'Color theme (simple, base16, ocean)')
@@ -65,24 +64,10 @@ program.on('--help', () => {

 program.parse(process.argv);

-if(program.android) {
-  program.platform = 'android';
-}
-
-if (program.linux) {
-  program.platform = 'linux';
-}
-
-if (program.osx) {
-  program.platform = 'osx';
-}
-
-if (program.sunos) {
-  program.platform = 'sunos';
-}
-
-if (program.windows) {
-  program.platform = 'windows';
+for (const platform in platforms.supportedPlatforms) {
+  if (program[platform]) {
+    program.platform = platform;
+  }
 }

 let cfg = config.get();
diff --git a/lib/platforms.js b/lib/platforms.js
index c938363..25f6a49 100644
--- a/lib/platforms.js
+++ b/lib/platforms.js
@@ -12,6 +12,8 @@ const folders = {
   'win32': 'windows'
 };

+const supportedPlatforms = Array.from(new Set(Object.values(folders))).sort();
+
 // Check if the platform is there in the list of platforms or not
 function isSupported(platform) {
   return Object.prototype.hasOwnProperty.call(folders, platform);
@@ -36,5 +38,6 @@ function getPreferredPlatformFolder(config) {
 module.exports = {
   isSupported,
   getPreferredPlatform,
-  getPreferredPlatformFolder
+  getPreferredPlatformFolder,
+  supportedPlatforms
 };
```

Now won't fill up `bin/tldr` with nearly identical if statements as new platforms are added.

File: bin/tldr
Comment (MasterOdin) [inline bin/tldr:29]: The list of options no longer matches the README / completion scripts, where now will see `darwin`, `macos`, and `win32` here, but are not completable.

Also, having them in any order looks weird, and would prefer to see them alphabetized here and also in the description for `--platform` so that it's easier to look for a specific platform:
```
  -p, --platform [type]    Override the current platform [osx, darwin, macos, linux, openbsd, sunos, android, windows,
                           win32]
  --osx                    Override the platform with osx
  --darwin                 Override the platform with darwin
  --macos                  Override the platform with macos
  --linux                  Override the platform with linux
  --openbsd                Override the platform with openbsd
  --sunos                  Override the platform with sunos
  --android                Override the platform with android
  --windows                Override the platform with windows
  --win32                  Override the platform with win32
```

Comment (kbdharun) [reply bin/tldr:29]: If possible, can you fix this alone? Directly, in this branch. Thanks in advance.

**Comment by kbdharun** [comment]
@MasterOdin I have rearranged the options in alphabetical order as requested.

## Changes in the last 3 commits

- Added support for FreeBSD and NetBSD platforms.
- Added the Client specification indication to the version's output.
- Updated website link and tags in `package.json`. 

In `package.json`, I have bumped the version to `3.4.0` as I plan on making a release after this PRs merge. With this the Node client is fully compliant with the latest client specification.

## Reference outputs

```zsh
➜  bin git:(feat/platform-option) ✗ ./tldr -v       
v3.3.8
Client Specification: 2.0
➜  bin git:(feat/platform-option) ✗ ./tldr -h
Usage: tldr|-o command [options]

Simplified and community-driven man pages

Options:
  -v, --version            Display version
  -l, --list               List all commands for the chosen platform in the cache
  -a, --list-all           List all commands in the cache
  -1, --single-column      List single command per line (use with options -l or -a)
  -r, --random             Show a random command
  -e, --random-example     Show a random example
  -f, --render [file]      Render a specific markdown [file]
  -m, --markdown           Output in markdown format
  -p, --platform [type]    Override the current platform [android, darwin, freebsd, linux, macos, netbsd, openbsd, osx, sunos,
                           win32, windows]
  --android                Override the platform with android
  --darwin                 Override the platform with darwin
  --freebsd                Override the platform with freebsd
  --linux                  Override the platform with linux
  --macos                  Override the platform with macos
  --netbsd                 Override the platform with netbsd
  --openbsd                Override the platform with openbsd
  --osx                    Override the platform with osx
  --sunos                  Override the platform with sunos
  --win32                  Override the platform with win32
  --windows                Override the platform with windows
  -t, --theme [theme]      Color theme (simple, base16, ocean)
  -s, --search [keywords]  Search pages using keywords
  -u, --update             Update the local cache
  -c, --clear-cache        Clear the local cache
  -h, --help               Show this help message

  Examples:

    $ tldr tar
    $ tldr du --platform=linux
    $ tldr --search "create symbolic link to file"
    $ tldr --list
    $ tldr --list-all
    $ tldr --random
    $ tldr --random-example

  To control the cache:

    $ tldr --update
    $ tldr --clear-cache

  To render a local file (for testing):

    $ tldr --render /path/to/file.md
```

**Comment by kbdharun** [comment]
## Changes in last commit

I used the 'he' library to decode HTML entities in email addresses, ensuring proper rendering. In it, I removed the "mailto:" prefix from email addresses, improving their presentation in the rendered content.

## Content

This fixes #395 where the GPG page would render with broken mail addresses like:

```md
➜  bin git:(feat/platform-option) ✗ ./tldr --render /home/kbdharunkrishna/Downloads/1.md

  gpg

  GNU Privacy Guard.
  See gpg2 for GNU Privacy Guard 2. Most operating systems symlink gpg to gpg2.
  More information: https://gnupg.org.

  - Create a GPG public and private key interactively:
    gpg --full-generate-key

  - Sign doc.txt without encryption (writes output to doc.txt.asc):
    gpg --clearsign doc.txt

  - Encrypt and sign doc.txt for mailto:&#x61;&#108;&#x69;&#x63;&#101;&#x40;&#x65;&#x78;&#x61;&#109;&#x70;&#108;&#101;&#46;&#x63;&#111;&#109; and mailto:&#98;&#x6f;&#98;&#x40;&#x65;&#x78;&#x61;&#109;&#112;&#x6c;&#x65;&#x2e;&#99;&#x6f;&#x6d; (output to doc.txt.gpg):
    gpg --encrypt --sign --recipient alice@example.com --recipient bob@example.com doc.txt

  - Encrypt doc.txt with only a passphrase (output to doc.txt.gpg):
    gpg --symmetric doc.txt

  - Decrypt doc.txt.gpg (output to stdout):
    gpg --decrypt doc.txt.gpg

  - Import a public key:
    gpg --import public.gpg

  - Export public key for mailto:&#x61;&#x6c;&#x69;&#x63;&#x65;&#x40;&#x65;&#120;&#x61;&#x6d;&#112;&#108;&#x65;&#x2e;&#99;&#x6f;&#x6d; (output to stdout):
    gpg --export --armor alice@example.com

  - Export private key for mailto:&#97;&#108;&#105;&#99;&#101;&#64;&#x65;&#120;&#x61;&#109;&#112;&#x6c;&#x65;&#x2e;&#99;&#111;&#109; (output to stdout):
    gpg --export-secret-keys --armor alice@example.com


See also: gpg2
```

Post the changes, it now renders like this:

```md
➜  bin git:(feat/platform-option) ✗ ./tldr --render /home/kbdharunkrishna/Downloads/1.md

  gpg

  GNU Privacy Guard.
  See gpg2 for GNU Privacy Guard 2. Most operating systems symlink gpg to gpg2.
  More information: https://gnupg.org.

  - Create a GPG public and private key interactively:
    gpg --full-generate-key

  - Sign doc.txt without encryption (writes output to doc.txt.asc):
    gpg --clearsign doc.txt

  - Encrypt and sign doc.txt for alice@example.com and bob@example.com (output to doc.txt.gpg):
    gpg --encrypt --sign --recipient alice@example.com --recipient bob@example.com doc.txt

  - Encrypt doc.txt with only a passphrase (output to doc.txt.gpg):
    gpg --symmetric doc.txt

  - Decrypt doc.txt.gpg (output to stdout):
    gpg --decrypt doc.txt.gpg

  - Import a public key:
    gpg --import public.gpg

  - Export public key for alice@example.com (output to stdout):
    gpg --export --armor alice@example.com

  - Export private key for alice@example.com (output to stdout):
    gpg --export-secret-keys --armor alice@example.com


See also: gpg2
```

**Comment by agnivade** [comment]
I am removing myself. Feel free to merge when you have 2 or more approvals.

**Comment by kbdharun** [comment]
> Do we have an ultimatum to merge this?

Waiting for @MasterOdin's review. After that is is GTG.

**Review by sbrl — COMMENTED** [review]
Lots of changes here, but a quick review doesn't reveal any major issues here. I suggest this is okay to merge. Stopping short of an approval though, since I'm not as involved in implementing the node client. /cc @MasterOdin.

**Comment by kbdharun** [comment]
@agnivade Sorry for the ping, I have a few doubts about how we publish a new release to the NPM (so that I can do it after this PR).

- I don't notice a release action (like the Python client) here. Would it automatically show up on NPM after creating a tagged release here? (If not can you add me to the Org over there as a collaborator [In this case, I will work on a release CI]).

**Comment by sbrl** [comment]
I think it would be helpful to have multiple people on the npm org, but unfortunately I don't have access. Might be worth checking who published recent versions of the Node client to see who to pester, @kbdharun?

**Comment by kbdharun** [comment]
> I think it would be helpful to have multiple people on the npm org, but unfortunately I don't have access. Might be worth checking who published recent versions of the Node client to see who to pester, @kbdharun?

I did the same 😅, after noticing the last release was made by @agnivade.

**Comment by ivanhercaz** [comment]
> how we publish a new release to NPM (so that I can do it after this PR)

Hi @kbdharun! As far I know, and according to [npm documentation](https://docs.npmjs.com/updating-your-published-package-version-number), to publish a new release you must change the version in `package.json` and then run `npm publish`, but I think that if the git repo is linked with the npm package, npm read the git tags and update it to the latest.

I will research more.

**Comment by agnivade** [comment]
Yes I release the npm package. Let me see if I can add you.

**Comment by kbdharun** [comment]
> Yes I release the npm package. Let me see if I can add you.

Thanks received the invite, accepted it. Can you do the same for tldr-lint too?

Kinda curious, don't we have all these packages inside a single Org on NPM?

**Comment by kbdharun** [comment]
**Update**: I have set up a new Org (https://www.npmjs.com/org/tldr-pages) in NPM. And added our existing non-scoped tldr package to it. While this isn't a package transfer, what is the benefit of the approach is we can set up new org-scoped packages in future and this allows adding collaborators via teams with ease.

-----

Already sent an invite to @agnivade, feel free to let me know your NPM username @sbrl  will add you too.

**Comment by kbdharun** [comment]
> > Do we have an ultimatum to merge this?
> 
> Waiting for @MasterOdin's review. After that is is GTG.

I think this PR is GTG now will merge it (after passing checks), the only unreviewed change is in `render.js` with mail URL decoding, other parts are cosmetic changes with reordering options (already reviewed and approved by Masterodin).

I have [reverted the initial suggestion to add `-o` and `--os` as .aliases](https://github.com/tldr-pages/tldr-node-client/pull/421#discussion_r1347487276) for backward compatibility as our current version of commander in `package.json` doesn't support it (updating dependencies with dependabot was problematic with us supporting Node versions 16 and above). IG we can have the newer options alone (as we would drop the older ones in future releases anyway.

**Pull request merged**
